vertx-sql-client/src/main/java/io/vertx/sqlclient/PreparedStatement.java

```diff
@@ -0,0 +1,80 @@
+/*
+ * Copyright (C) 2017 Julien Viet
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ *
+ */
+
+package io.vertx.sqlclient;
+
+import io.vertx.core.Future;
+import io.vertx.sqlclient.impl.ArrayTuple;
+import io.vertx.codegen.annotations.VertxGen;
+import io.vertx.core.AsyncResult;
+import io.vertx.core.Handler;
+
+/**
+ * A prepared statement, the statement is pre-compiled and
+ * it's more efficient to execute the statement for multiple times.
+ * In addition, this kind of statement provides protection against SQL injection attacks.
+ *
+ * @author <a href="mailto:julien@julienviet.com">Julien Viet</a>
+ */
+@VertxGen
+public interface PreparedStatement {
+
+  /**
+   * Create a query.
+   *
+   * @return the query
+   */
+  Query<RowSet<Row>> query();
+
+  /**
+   * @return create a query cursor with a {@code fetch} size and empty arguments
+   */
+  default Cursor cursor() {
+    return cursor(ArrayTuple.EMPTY);
+  }
+
+  /**
+   * Create a cursor with the provided {@code arguments}.
+   *
+   * @param args the list of arguments
+   * @return the query
+   */
+  Cursor cursor(Tuple args);
+
+  /**
+   * Execute the prepared query with a cursor and createStream the result. The createStream opens a cursor
+   * with a {@code fetch} size to fetch the results.
+   * <p/>
+   * Note: this requires to be in a transaction, since cursors require it.
+   *
+   * @param fetch the cursor fetch size
+   * @param args the prepared query arguments
+   * @return the createStream
+   */
+  RowStream<Row> createStream(int fetch, Tuple args);
```

Comment:
Since `cursor()` and `createStream()` are both different flavors of executing queries, it seems that they would be better suited on the `Query` interface. 

Basically there are 4 ways that a user can retrieve results from their query:
 - RowSet<Row> (default way)
 - Function<Row,U>
 - Collector
 - Cursor (prepared statement only)
 - RowStream<Row> (prepared statement only)

I think it would be good if we only had 1 way of executing a query (via `execute` on the `Query` interface) and the result type could be selected prior to execution. For example:
 - (nothing needed to get the default RowSet)
 - `Query#mapping(Function<Row,U>)`
 - `Query#collecting(Collector)`
 - `Query#streaming(int fetchSize)` (proposed)
 - `Query#withCursor()` (proposed)

Reply:
Cursor and stream should be kept away from `Query` as they can require transaction isolation (in PG at least) and they should be handled only from `SqlConnection` and `PreparedStatement` and are not one shot operations (i.e they span several interactions). 

Providing cursor on `Query` would make it available from a pool without acquiring a connection.

Reply:
> Providing cursor on Query would make it available from a pool without acquiring a connection.

I was suggesting that the proposed `streaming` and `withCursor()` methods would alter the result type once `execute` is eventually invoked, so it would still require a connection. Example usage might look like:
```java
pool.createPreparedQuery(SQL)
       .withCursor()
       .execute(ar -> { 
         Cursor c = ar.result();
         // do stuff with cursor
       });
```
So this way `execute` would be the only terminal operation in the API (for executing queries). Otherwise we have 3 terminal operations (execute, cursor, and stream).

Reply:
As I said, using this from a pool is error prone. A cursor is an advanced feature and the user should use a connection to interact with a Cursor and it does not fit well with the `SqlClient` interface that provides exclusively one-shot operation that do not require any close operation to release the resource.